synapse/storage/schema/main/delta/78/03_add_unused_expires_at_for_media.sql

```diff
@@ -0,0 +1,20 @@
+/* Copyright 2023 Beeper Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+-- Add new colums to the `local_media_repository` to keep track of when the
+-- media ID must be used by. This is to support async uploads (see MSC2246).
+
+ALTER TABLE local_media_repository
+  ADD COLUMN unused_expires_at BIGINT DEFAULT NULL;
```

Comment:
We might need to add an index here to make it fast like we had to to do for `/timestamp_to_event`, https://github.com/matrix-org/synapse/pull/14799/files

Reply:
Would I need to do it as a background task like `events_bg_updates` does?

Reply:
Do we even need an index? (I was just making an educated guess based on previous experience)

> do it as a background task like events_bg_updates does?

:+1: Example https://github.com/matrix-org/synapse/pull/14799/files

Reply:
![image](https://github.com/matrix-org/synapse/assets/16734772/fea39c3e-eb03-414f-9d11-c6925c8231de)

I can't create the index concurrently, which means that it will probably take forever to create this index and block startup, right?

Reply:
We probably only need an index if we're frequently looking things up by `unused_expires_at`, I don't see any code that does this so we likely don't need one?

----

Regardless, the way to do this would be to create it in the background by doing something like this:

https://github.com/matrix-org/synapse/blob/2ad91ec628126753590c1a90c432270d6c8fa8fd/synapse/storage/databases/main/event_push_actions.py#L284-L290

And then adding a background job in the schema file, something like:

https://github.com/matrix-org/synapse/blob/666ae877292d4747b9441105e3df8558f7a335c0/synapse/storage/schema/main/delta/72/06thread_notifications.sql#L26-L27

Reply:
I'm using the `unused_expires_at` in a query here: https://github.com/matrix-org/synapse/pull/15503/files#diff-f05c2d0a285e2b60b93cbffe6283864c9b23c19984fe7e4511604797413acc06R424

Reply:
But you're also selecting on `user_id`, which *does* have an index already. (The index is `user_id, created_ts`.)

Reply:
Is a linear scan on the user's entire list of media going to be fast though?

Reply:
It depends on how much media the user has, I think. Would it make more sense to re-use that index and change all the conditions to be based on `created_ts` instead of a new expiration time? E.g.

```sql
SELECT COUNT(*), MIN(created_ts)
FROM local_media_repository
WHERE user_id = ?
	AND created_ts > ?
	AND media_length IS NULL
)
```

And then provide the arguments such that you do `self._clock.time_msec() - unused_expiration_time`.

Reply:
Updated this PR to do expirations based on `created_ts`. If in the future we want to have variable expirations, we will need to invent something smart here, but for now this is simple enough.